### ISSUE TITLE
Reproducibility of Table 3

### ISSUE THREAD

**Issue by SilvioGiancola** [inyeoplee77/SAGPool#5]
Hi, 

I am running your code several time on the DD dataset and obtain different results than the one you present on Table 3 of your paper. 
In particular I run 20 times this experiment, estimate the average and std but find out that the training is very random, with the std rising up to +-10%. Note that I used the same hyperparameters you provide in your paper and your google sheet (see [issue #2](https://github.com/inyeoplee77/SAGPool/issues/2#issuecomment-502968633)).
I also tried with the [ReduceLROnPlateau](https://pytorch.org/docs/stable/optim.html#torch.optim.lr_scheduler.ReduceLROnPlateau) scheduler for the LR, but still have an std up to 5%.
How did you select your seed and how come there is such variation?

Thank you for your support,

Best,

**Comment by ThyrixYang** [comment]
I can confirm with @SilvioGiancola that the variance is much larger than results on google sheet, by using hyperparameters as suggested on the dataset D&D. And the mean accuracy seems not that high.

**Comment by ThyrixYang** [comment]
Hi @SilvioGiancola , they said in their paper that
> In our experiments, we evaluated the pooling methods over 20 random seeds using 10-fold cross validation. A total of 200 testing results were used to obtain the final accuracy of each method on each dataset

So we evaluate by run CV (the main.py file) 10 times, calculate their mean, take this as **one** result, and repeat this procedure 20 times. 
I hope this helps.

**Comment by SilvioGiancola** [comment]
Hi @ThyrixYang , thank you for sharing this detail!

In that case, it's not *exactly* like running 10 times the code as the main.py is not doing 10-fold cross validation but [random splitting](https://github.com/inyeoplee77/SAGPool/blob/4988acddc9319681826fb03fe6141701e9714693/main.py#L48).

With this averaging over 10 runs, did you get a variance similar than one in Table 3?

**Comment by ThyrixYang** [comment]
@SilvioGiancola yes, the variance is similar, although its mean is a bit lower.

**Comment by SilvioGiancola** [comment]
@ThyrixYang I solved the variance issue with this 10-fold cross validation. 

Although, when I reproduce their results, I am getting 10% lower than what they claim on the DD dataset, using the global pooling model. Are you also having such a big difference in your results?

I wished the authors could provide a code to reproduce their results. It is impossible to build upon them...

**Comment by ThyrixYang** [comment]
@SilvioGiancola Are you doing exactly 10-fold CV? 
I remember on the random CV that it's about 2~3% lower, not 10%
I'm working on a paper about graph pooling now, but not based mainly on this paper, I will do more experiments later, maybe we can share some results then.

**Comment by SilvioGiancola** [comment]
@ThyrixYang I'll be more than happy to share some results with you on this baseline.
I guess I am performing the 10-fold CV properly, in particular:

1. I randomly split the dataset (DD in my case) in 10 folds of same length (last fold has a slight different length)
2. I use 9 folds for training, the 10th for testing
3. I repeat the 9-fold training 10 times, segregating a different fold for the testing at each time
4. I average the testing performance over the 10 folds -> this gives me the results for 1 run
5. I repeat steps 1-4 for 20 times, with 20 different random splits for the folds
6. I estimate the average and the std over the 20 runs

I get `65.1 ± 1.23` on DD using the global pooling setting, while the paper claim `76.19 ± 0.94`. BTW I tried both the SAGPool implementations from this repo and from [pytorch-geometric](https://github.com/rusty1s/pytorch_geometric), with similar results. I also used the hyper parameters from the [gsheet](https://docs.google.com/spreadsheets/d/1JXGNOCQkRHDCQqNarteYpEuWnkNzNq_WFiQrIY276i0/edit#gid=0) (lr=0.005, nhid=128, weight decay=0.00001)

Are you doing anything different for the 10 fold CV? Have you tried the same dataset or a different one?

**Comment by ThyrixYang** [comment]
Here is my result on DD
```
{'batch_size': 128, 'lr': 0.005, 'nhid': 128, 'weight_decay': 0.0001, 'pooling_ratio': 0.5, 'dropout_ratio': 0.5, 'dataset': 'DD', 'epochs': 100000, 'patience': 50, 'pooling_layer_type': 'GCNConv', 'seed': 138004, 'num_classes': 2, 'num_features': 89}
mean accuarcy 0.8050420168067227, std: 0.007325880577379257

{'batch_size': 128, 'lr': 0.005, 'nhid': 128, 'weight_decay': 0.0001, 'pooling_ratio': 0.5, 'dropout_ratio': 0.5, 'dataset': 'DD', 'epochs': 100000, 'patience': 50, 'pooling_layer_type': 'GCNConv', 'seed': 639129, 'num_classes': 2, 'num_features': 89}
mean accuarcy 0.7571428571428571, std: 0.01214859856705964

{'batch_size': 128, 'lr': 0.005, 'nhid': 128, 'weight_decay': 0.0001, 'pooling_ratio': 0.5, 'dropout_ratio': 0.5, 'dataset': 'DD', 'epochs': 100000, 'patience': 50, 'pooling_layer_type': 'GCNConv', 'seed': 255038, 'num_classes': 2, 'num_features': 89}
mean accuarcy 0.7571428571428571, std: 0.006980356187326114

{'batch_size': 128, 'lr': 0.005, 'nhid': 128, 'weight_decay': 0.0001, 'pooling_ratio': 0.5, 'dropout_ratio': 0.5, 'dataset': 'DD', 'epochs': 100000, 'patience': 50, 'pooling_layer_type': 'GCNConv', 'seed': 128397, 'num_classes': 2, 'num_features': 89}
mean accuarcy 0.7285714285714286, std: 0.01764705882352942

{'batch_size': 128, 'lr': 0.005, 'nhid': 128, 'weight_decay': 0.0001, 'pooling_ratio': 0.5, 'dropout_ratio': 0.5, 'dataset': 'DD', 'epochs': 100000, 'patience': 50, 'pooling_layer_type': 'GCNConv', 'seed': 433605, 'num_classes': 2, 'num_features': 89}
mean accuarcy 0.788235294117647, std: 0.016294722209802805

{'batch_size': 128, 'lr': 0.005, 'nhid': 128, 'weight_decay': 0.0001, 'pooling_ratio': 0.5, 'dropout_ratio': 0.5, 'dataset': 'DD', 'epochs': 100000, 'patience': 50, 'pooling_layer_type': 'GCNConv', 'seed': 255803, 'num_classes': 2, 'num_features': 89}
mean accuarcy 0.7722689075630251, std: 0.024772105836947365

{'batch_size': 128, 'lr': 0.005, 'nhid': 128, 'weight_decay': 0.0001, 'pooling_ratio': 0.5, 'dropout_ratio': 0.5, 'dataset': 'DD', 'epochs': 100000, 'patience': 50, 'pooling_layer_type': 'GCNConv', 'seed': 745153, 'num_classes': 2, 'num_features': 89}
mean accuarcy 0.753781512605042, std: 0.011305566426112343

{'batch_size': 128, 'lr': 0.005, 'nhid': 128, 'weight_decay': 0.0001, 'pooling_ratio': 0.5, 'dropout_ratio': 0.5, 'dataset': 'DD', 'epochs': 100000, 'patience': 50, 'pooling_layer_type': 'GCNConv', 'seed': 726127, 'num_classes': 2, 'num_features': 89}
mean accuarcy 0.7243697478991598, std: 0.021192470946081485

{'batch_size': 128, 'lr': 0.005, 'nhid': 128, 'weight_decay': 0.0001, 'pooling_ratio': 0.5, 'dropout_ratio': 0.5, 'dataset': 'DD', 'epochs': 100000, 'patience': 50, 'pooling_layer_type': 'GCNConv', 'seed': 949864, 'num_classes': 2, 'num_features': 89}
mean accuarcy 0.8445378151260504, std: 0.012031782406114558

{'batch_size': 128, 'lr': 0.005, 'nhid': 128, 'weight_decay': 0.0001, 'pooling_ratio': 0.5, 'dropout_ratio': 0.5, 'dataset': 'DD', 'epochs': 100000, 'patience': 50, 'pooling_layer_type': 'GCNConv', 'seed': 754063, 'num_classes': 2, 'num_features': 89}
mean accuarcy 0.76890756302521, std: 0.007747516350666312

{'batch_size': 128, 'lr': 0.005, 'nhid': 128, 'weight_decay': 0.0001, 'pooling_ratio': 0.5, 'dropout_ratio': 0.5, 'dataset': 'DD', 'epochs': 100000, 'patience': 50, 'pooling_layer_type': 'GCNConv', 'seed': 851503, 'num_classes': 2, 'num_features': 89}
mean accuarcy 0.7033613445378151, std: 0.015055859552242786
```

**Comment by jiaruHithub** [comment]
> @ThyrixYang I'll be more than happy to share some results with you on this baseline.
> I guess I am performing the 10-fold CV properly, in particular:
> 
> 1. I randomly split the dataset (DD in my case) in 10 folds of same length (last fold has a slight different length)
> 2. I use 9 folds for training, the 10th for testing
> 3. I repeat the 9-fold training 10 times, segregating a different fold for the testing at each time
> 4. I average the testing performance over the 10 folds -> this gives me the results for 1 run
> 5. I repeat steps 1-4 for 20 times, with 20 different random splits for the folds
> 6. I estimate the average and the std over the 20 runs
> 
> I get `65.1 ± 1.23` on DD using the global pooling setting, while the paper claim `76.19 ± 0.94`. BTW I tried both the SAGPool implementations from this repo and from [pytorch-geometric](https://github.com/rusty1s/pytorch_geometric), with similar results. I also used the hyper parameters from the [gsheet](https://docs.google.com/spreadsheets/d/1JXGNOCQkRHDCQqNarteYpEuWnkNzNq_WFiQrIY276i0/edit#gid=0) (lr=0.005, nhid=128, weight decay=0.00001)
> 
> Are you doing anything different for the 10 fold CV? Have you tried the same dataset or a different one?

Hi, I tried to reproduce the experiment too.
How is your replication of the experiment now?
I am looking forward to your reply. I am very interested in it
Thank you!

**Comment by Abelpzx** [comment]
Hi, each time I run this code, I got different results.  I have set these seeds but still got different results.
 
    torch.cuda.manual_seed(12345)  
    torch.cuda.manual_seed_all(12345)  
    random.seed(12345)
    np.random.seed(12345)
    #torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False

What can I do to get the sample after running this code each time?
I am looking forward to your reply. 
Thank you!
